subsys/net/lib/shell/dhcpv4.c

```diff
@@ -0,0 +1,206 @@
+		return UINT32_MAX;
+	}
+
+	return k_ticks_to_ms_floor64(timeout.ticks) / 1000;
+}
+
+static void dhcpv4_lease_cb(struct net_if *iface,
+			    struct dhcpv4_addr_slot *lease,
+			    void *user_data)
+{
+	struct net_shell_user_data *data = user_data;
+	const struct shell *sh = data->sh;
+	int *count = data->user_data;
+	char expiry_str[] = "4294967295"; /* Lease time is uint32_t, so take
+					   * theoretical max.
+					   */
+
+	if (*count == 0) {
+		PR("            Address\t    State\tExpiry (sec)\n");
+	}
+
+	(*count)++;
+
+	if (lease->state == DHCPV4_SERVER_ADDR_DECLINED) {
+		snprintk(expiry_str, sizeof(expiry_str) - 1, "infinite");
+	} else {
+		snprintk(expiry_str, sizeof(expiry_str) - 1, "%u",
+			 timepoint_to_s(lease->expiry));
+	}
+
+	PR("%2d. %15s\t%9s\t%12s\n", *count, net_sprint_ipv4_addr(&lease->addr),
+	   dhcpv4_addr_state_to_str(lease->state), expiry_str);
+
+}
+#endif /* CONFIG_NET_DHCPV4_SERVER */
+
+static int cmd_net_dhcpv4_server_status(const struct shell *sh, size_t argc, char *argv[])
+{
+#if defined(CONFIG_NET_DHCPV4_SERVER)
+	struct net_shell_user_data user_data;
+	struct net_if *iface = NULL;
+	int idx, ret;
+	int count = 0;
+
+	idx = get_iface_idx(sh, argv[1]);
+	if (idx < 0) {
+		return -ENOEXEC;
+	}
+
+	iface = net_if_get_by_index(idx);
+	if (!iface) {
+		PR_WARNING("No such interface in index %d\n", idx);
+		return -ENOEXEC;
+	}
+
+	user_data.sh = sh;
+	user_data.user_data = &count;
+
+	ret = net_dhcpv4_server_foreach_lease(iface, dhcpv4_lease_cb, &user_data);
+	if (ret == -ENOENT) {
+		PR_WARNING("DHCPv4 server is not running on interface %d\n", idx);
+	} else if (count == 0) {
+		PR("DHCPv4 server - no addresses assigned on interface %d\n", idx);
+	}
+#else /* CONFIG_NET_DHCPV4_SERVER */
+	PR_INFO("Set CONFIG_NET_DHCPV4_SERVER to enable DHCPv4 server support.\n");
+#endif /* CONFIG_NET_DHCPV4_SERVER */
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(net_cmd_dhcpv4_server,
+	SHELL_CMD_ARG(start, NULL, "Start the DHCPv4 server operation on the interface.\n"
+		      "'net dhcpv4 server start <index> <base address>'\n"
+		      "<index> is the network interface index.\n"
+		      "<base address> is the first address for the address pool.",
+		      cmd_net_dhcpv4_server_start, 3, 0),
+	SHELL_CMD_ARG(stop, NULL, "Stop the DHCPv4 server operation on the interface.\n"
+		      "'net dhcpv4 server stop <index>'\n"
+		      "<index> is the network interface index.",
+		      cmd_net_dhcpv4_server_stop, 2, 0),
+	SHELL_CMD_ARG(status, NULL, "Print the DHCPv4 server status on the interface.\n"
+		      "'net dhcpv4 server status <index>'\n"
+		      "<index> is the network interface index.",
+		      cmd_net_dhcpv4_server_status, 2, 0),
+	SHELL_SUBCMD_SET_END
+);
+
+SHELL_STATIC_SUBCMD_SET_CREATE(net_cmd_dhcpv4,
+	SHELL_CMD(server, &net_cmd_dhcpv4_server,
+		  "DHCPv4 server service management.",
+		  NULL),
+	SHELL_SUBCMD_SET_END
+);
+
+SHELL_SUBCMD_ADD((net), dhcpv4, &net_cmd_dhcpv4, "Manage DHPCv4 services.",
```

Comment:
Should we have also some commands for the DHCPv4 client? I recently noticed that if I enable DHCPv4 but disable config module, then I need to add a call to `net_dhcpv4_start/stop` to some file. For debugging purposes it would be useful to support the client side in this shell. Naturally this could be in another PR.

Reply:
Yes, I thought it'd be useful to have commands for DHCP client as well, didn't want to mix them in here with the server specific stuff. I can prepare a PR with new shell commands and the source file relocation once we're done with the server.